### PR TITLE
C-file

### PR DESCRIPTION
**Pull request by heifner**
- Wrapper for c-file access that provides a similar interface as `fstream` without all the overhead of std streams.

### CODE REVIEW

File: include/fc/io/cfile.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+   size_t file_size() const {
+      return fc::file_size( _file_path );
+   }
+
+   /// @param mode is any mode supported by fopen
+   ///        Tested with:
+   ///         "ab+" - open for binary update - create if does not exist
+   ///         "rb+" - open for binary update - file must exist
+   void open( const char* mode ) {
+      _file.reset( fopen( _file_path.generic_string().c_str(), mode ) );
```
Comment (spoonincode) [inline include/fc/io/cfile.hpp]: fopen() has some weirdness in win32 if wanting to support unicode filenames. See EOSIO/eos#7640 for what I did in the only other place we've used fopen. Not sure if better solution exists

Comment (heifner) [reply include/fc/io/cfile.hpp]: Added similar fix as #7640

File: include/fc/io/cfile.hpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      _file_path = std::move( file_path );
+   }
+
+   fc::path get_file_path() const {
+      return _file_path;
+   }
+
+   bool is_open() const { return _open; }
+
+   size_t file_size() const {
```
Comment (jgiszczak) [inline include/fc/io/cfile.hpp]: stat() of an open file can result in unexpected behavior, up to and including loss of the contents of the file.  Consider requiring the file be closed for file_size to work.

File: include/fc/io/cfile.hpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+   void close() {
+      _file.reset();
+      _open = false;
+   }
+
+   void remove() {
+      if( _open ) {
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() + " Unable to remove as file is open" );
+      }
+      fc::remove_all( _file_path );
```
Comment (jgiszczak) [inline include/fc/io/cfile.hpp]: This call as well as `fc::file_size()` above are one-liner wrappers around the `boost::filesystem` calls of the same name, using the boost API that can throw `filesystem_error`.  Presumably `boost::filesystem_error`, though boost's documentation doesn't specify.  `fc` does not catch any exception.

File: test/io/test_cfile.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+      int x = 42, y = 0;
+      t.seek( 1 );
+      t.write( reinterpret_cast<char*>( &x ), sizeof( x ) );
+      t.seek( 1 );
+      t.read( reinterpret_cast<char*>( &y ), sizeof( y ) );
+      BOOST_CHECK_EQUAL( x, y );
+
+      t.close();
+      BOOST_CHECK( !t.is_open() );
```
[the remark this answers is not in the document]
Comment (heifner) [reply test/io/test_cfile.cpp]: Added.

File: include/fc/io/cfile.hpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      size_t result = fwrite( d, 1, n, _file.get() );
+      if( result != n ) {
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() +
+                                   " unable to write " + std::to_string( n ) + " only wrote " + std::to_string( result ) );
+      }
+   }
+
+   void flush() {
+      if( 0 != fflush( _file.get() ) ) {
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() + " unable to flush file." );
```
[the remark this answers is not in the document]
Comment (heifner) [reply include/fc/io/cfile.hpp]: Added ferror result to exception message.

File: include/fc/io/cfile.hpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() +
+                                   " unable to read " + std::to_string( n ) + " only read " + std::to_string( result ) );
+      }
+   }
+
+   void write( const char* d, size_t n ) {
+      size_t result = fwrite( d, 1, n, _file.get() );
+      if( result != n ) {
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() +
+                                   " unable to write " + std::to_string( n ) + " only wrote " + std::to_string( result ) );
```
[the remark this answers is not in the document]
Comment (heifner) [reply include/fc/io/cfile.hpp]: Added.

File: include/fc/io/cfile.hpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() +
+                                   " unable to SEEK_END to: " + std::to_string(loc) );
+      }
+   }
+
+   void read( char* d, size_t n ) {
+      size_t result = fread( d, 1, n, _file.get() );
+      if( result != n ) {
+         throw std::runtime_error( "cfile: " + _file_path.generic_string() +
+                                   " unable to read " + std::to_string( n ) + " only read " + std::to_string( result ) );
```
[the remark this answers is not in the document]
Comment (heifner) [reply include/fc/io/cfile.hpp]: Added.

File: include/fc/io/cfile.hpp (hunk 2 above)
Comment (heifner) [reply include/fc/io/cfile.hpp]: We were actually using it in block_log on open files. I'll fix that on the PR to incorporate this into block_log.cpp.

File: include/fc/io/cfile.hpp (hunk 3 above)
Comment (heifner) [reply include/fc/io/cfile.hpp]: Good point. I've removed `remove` and `file_size` and will just use the existing `fc` calls directly.

**Pull request merged**